nexus/src/app/background/tasks/support_bundle_collector.rs

```diff
@@ -0,0 +1,1450 @@
+            .await?;
+
+        // Returning from this method should drop all temporary storage
+        // allocated locally for this support bundle.
+        Ok(report)
+    }
+
+    // Perform the work of collecting the support bundle into a temporary directory
+    //
+    // - "dir" is a directory where data can be stored.
+    // - "bundle" is metadata about the bundle being collected.
+    //
+    // If a partial bundle can be collected, it should be returned as
+    // an Ok(CollectionReport). Any failures from this function will prevent
+    // the support bundle from being collected altogether.
+    //
+    // NOTE: The background task infrastructure will periodically check to see
+    // if the bundle has been cancelled by a user while it is being collected.
+    // If that happens, this function will be CANCELLED at an await point.
+    //
+    // As a result, it is important that this function be implemented as
+    // cancel-safe.
+    async fn collect_bundle_as_file(
+        &self,
+        dir: &Utf8TempDir,
+    ) -> anyhow::Result<CollectionReport> {
+        let log = &self.log;
+
+        info!(&log, "Collecting bundle as local file");
+        let mut report = CollectionReport::new(self.bundle.id.into());
+
+        tokio::fs::write(
+            dir.path().join("bundle_id.txt"),
+            self.bundle.id.to_string(),
+        )
+        .await?;
+
+        if let Ok(all_sleds) = self
+            .collector
+            .datastore
+            .sled_list_all_batched(&self.opctx, SledFilter::InService)
+            .await
+        {
+            report.listed_in_service_sleds = true;
+
+            // NOTE: This could be, and probably should be, done concurrently.
+            for sled in &all_sleds {
+                info!(&log, "Collecting bundle info from sled"; "sled" => %sled.id());
+                let sled_path = dir
+                    .path()
+                    .join("rack")
+                    .join(sled.rack_id.to_string())
+                    .join("sled")
+                    .join(sled.id().to_string());
+                tokio::fs::create_dir_all(&sled_path).await?;
+                tokio::fs::write(
+                    sled_path.join("sled.txt"),
+                    format!("{sled:?}"),
+                )
+                .await?;
+
+                if self.request.skip_sled_info {
+                    continue;
+                }
+
+                let Ok(sled_client) = nexus_networking::sled_client(
+                    &self.collector.datastore,
+                    &self.opctx,
+                    sled.id(),
+                    log,
+                )
+                .await
+                else {
+                    tokio::fs::write(
+                        sled_path.join("error.txt"),
+                        "Could not contact sled",
+                    )
+                    .await?;
+                    continue;
+                };
+
+                write_command_result_or_error(
```

Comment:
Unrelated to this PR but I wish we had something like `#[must_use]` for the sled-agent things being added that told you to come here and add them to the collection.

Reply:
Maybe we can talk about this in the support bundle sync? Are you trying to trigger a compiler error in Nexus if a Sled Agent API hasn't been called here?

It might be easier to make a test for this behavior, rather than trying to make it a compile error.